webauthn-rs-core/src/core.rs

```diff
@@ -843,87 +786,37 @@ impl WebauthnCore {
-        if policies.len() > 1 {
-            return Err(WebauthnError::InconsistentUserVerificationPolicy);
-        }
-        let policy = policies
-            .pop_first()
-            .ok_or(WebauthnError::CredentialNotFound)?;
-
-        self.generate_challenge_authenticate_inner(creds, policy, extensions, false)
-    }
-
-    /// Authenticate a single credential, with the ability to override the userVerification
-    /// policy requested, or extensions in use. If userVerification is `None`, the policy from
-    /// registration is used.
-    ///
-    /// NOTE: Over-riding the UserVerificationPolicy may have SECURITY consequences. You should
-    /// understand how this interacts with the single credential in use, and how that may impact
-    /// your system security.
-    ///
-    /// If in doubt, do NOT use this function!
-    pub fn generate_challenge_authenticate_credential(
-        &self,
-        cred: Credential,
-        policy: Option<UserVerificationPolicy>,
-        extensions: Option<RequestAuthenticationExtensions>,
-    ) -> Result<(RequestChallengeResponse, AuthenticationState), WebauthnError> {
-        let policy = policy.unwrap_or(cred.registration_policy);
-        self.generate_challenge_authenticate_inner(vec![cred], policy, extensions, false)
-    }
-
     /// Authenticate a set of credentials allowing the user verification policy to be set.
+    /// If no credentials are provided, this will start a discoverable credential authentication.
     ///
-    /// NOTE: Over-riding the UserVerificationPolicy may have SECURITY consequences. You should
-    /// understand how this interacts with the single credential in use, and how that may impact
-    /// your system security.
+    /// Policy defines the require UserVerification policy. This MUST be consistent with the
+    /// credentials being used in the authentication.
     ///
-    /// If in doubt, do NOT use this function!
-    pub fn generate_challenge_authenticate_policy(
+    /// `allow_backup_eligible_upgrade` allows rejecting credentials whos backup eligibility
+    /// has changed between registration and authentication.
+    pub fn generate_challenge_authenticate(
         &self,
         creds: Vec<Credential>,
-        policy: UserVerificationPolicy,
+        policy: Option<UserVerificationPolicy>,
         extensions: Option<RequestAuthenticationExtensions>,
-        allow_backup_eligible_upgrade: bool,
+        allow_backup_eligible_upgrade: Option<bool>,
     ) -> Result<(RequestChallengeResponse, AuthenticationState), WebauthnError> {
-        self.generate_challenge_authenticate_inner(
-            creds,
-            policy,
-            extensions,
-            allow_backup_eligible_upgrade,
-        )
-    }
+        let policy = if let Some(policy) = policy {
+            policy
+        } else {
+            let mut policies =
+                BTreeSet::from_iter(creds.iter().map(|cred| cred.registration_policy.to_owned()));
```

Comment:
This creates a copy of all `creds` registration policies, and then deduplicates them, and then only actually passes through the first policy.

This could instead just copy the first registration policy it sees, then checks that all subsequent policies are the same. If one differs, break immediately with `InconsistentUserVerificationPolicy`.

If `creds` is empty, then break with `CredentialNotFound`.